modules/docs/_util.nix

```diff
@@ -0,0 +1,166 @@
+    {
+      options = {
+        enable = lib.mkOption {
+          type = lib.types.bool;
+          description = "Whether to enable this page/menu item.";
+          default = true;
+          example = false;
+        };
+        target = lib.mkOption {
+          type = lib.types.str;
+          description = ''
+            The target filepath, relative to the root of the docs.
+          '';
+          default = lib.optionalString (name != "") (name + "/") + "index.md";
+          defaultText = lib.literalMD ''
+            `<name>` joined with `"index.md"`. Separated by `"/"` if `<name>` is non-empty.
+          '';
+        };
+        source = lib.mkOption {
+          type = with lib.types; nullOr path;
+          description = ''
+            Markdown page. Set to null to create a menu entry without a corresponding file.
+          '';
+        };
+        menu.location = lib.mkOption {
+          type = with lib.types; listOf str;
+          description = ''
+            A location path that represents the page's position in the menu tree.
+
+            The text displayed in the menu is derived from this value,
+            after the location of any parent nodes in the tree is removed.
+
+            For example, if this page has the location `[ "foo" "bar" ]`
+            and there is another page with the location `[ "foo" ]`,
+            then the menu will render as:
+            ```markdown
+            - foo
+              - bar
+            ```
+
+            However if there was no other page with the `[ "foo" ]` location,
+            the menu would instead render as:
+            ```markdown
+            - foo.bar
+            ```
+          '';
+          default =
+            let
+              list = lib.splitString "/" config.target;
+              last = lib.last list;
+              rest = lib.dropEnd 1 list;
+            in
+            if last == "index.md" then
+              rest
+            else if lib.hasSuffix ".md" last then
+              rest ++ [ (lib.removeSuffix ".md" last) ]
+            else
+              list;
+          defaultText = lib.literalMD ''
+            `target`, split by `"/"`, with any trailing `"index.md` or `".md"` suffixes removed.
+          '';
+        };
+        menu.section = lib.mkOption {
+          type = lib.types.enum (builtins.attrNames sections);
+          description = ''
+            Determines the menu section.
+
+            Must be a section defined in `docs.menu.sections`.
+          '';
+        };
+      };
+    };
+in
+{
+  options.docs._utils = lib.mkOption {
+    type = with lib.types; lazyAttrsOf raw;
+    description = "internal utils, modules, functions, etc";
+    default = { };
+    internal = true;
+    visible = false;
+  };
+
+  config.docs._utils = {
+    # A liberal type that permits any superset of docsPageModule
+    docsPageLiberalType = lib.types.submodule [
+      { _module.check = false; }
+      docsPageModule
+    ];
```

Comment:
What is the difference between this and a freeform option?

Reply:
Ah this is an "extendable" submodule in some sense?

Reply:
A freeform submodule allows `config` attrs to be defined in a freeform way.

On the other hand, this "liberal" submodule just discards any definitions that don't correspond to a declared option.

I.e. option existence checks are disabled.